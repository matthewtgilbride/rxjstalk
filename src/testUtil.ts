import { TestScheduler } from "rxjs/testing";


export const objectEqualityTestScheduler = () => new TestScheduler((actual, expected) => {
  // asserting the two objects are equal
  expect(actual).toEqual(expected);
});

export const snapshotTestScheduler = () => new TestScheduler((actual) => {
  // asserting saving to snapshot
  expect(actual).toMatchSnapshot()
});
