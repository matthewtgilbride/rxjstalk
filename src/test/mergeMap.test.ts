import { objectEqualityTestScheduler, snapshotTestScheduler } from "../testUtil";
import { mergeMap, map } from "rxjs/operators";


describe('mergeMap', () => {

  it('plain mergeMap', () => {
    objectEqualityTestScheduler().run(({ cold, expectObservable}) => {

      const in1 = 'abc --- --- --- |'
      const in2 = '1-- 2-- 3-- |'

      const out = 'abc def ghi --- |'

      const values = {
        a: 'a1',
        b: 'b1',
        c: 'c1',
        d: 'a2',
        e: 'b2',
        f: 'c2',
        g: 'a3',
        h: 'b3',
        i: 'c3'
      }

      const in1$ = cold(in1)
      const in2$ = cold(in2)

      const outputStream = in1$.pipe(
        mergeMap((x) => in2$.pipe(map(y => `${x}${y}`))),
      )

      expectObservable(outputStream).toBe(out, values)

    })
  })

  it('snapshot', () => {
    snapshotTestScheduler().run(({cold, expectObservable}) => {

      const in1 = 'abc --- --- --- --- |'
      const in2 = '1-- 2-- 3-- |'


      const in1$ = cold(in1)
      const in2$ = cold(in2)

      const outputStream = in1$.pipe(
        mergeMap(x => in2$.pipe(map(y => x + y)), 2)
      )

      expectObservable(outputStream).toBe('')

    })
  })

  it('equality', () => {
    objectEqualityTestScheduler().run(({cold, expectObservable}) => {

      const in1 = 'abc --- --- --- --- |'
      const in2 = '1-- 2-- 3-- |'
      const out = 'qr- st- uv- x-- y-- z-- |'

      const values = {
        q: 'a1',
        r: 'b1',
        s: 'a2',
        t: 'b2',
        u: 'a3',
        v: 'b3',
        x: 'c1',
        y: 'c2',
        z: 'c3',
      }


      const in1$ = cold(in1)
      const in2$ = cold(in2)

      const outputStream = in1$.pipe(
        mergeMap(x => in2$.pipe(map(y => x + y)), 2)
      )

      expectObservable(outputStream).toBe(out, values)

    })
  })
})
