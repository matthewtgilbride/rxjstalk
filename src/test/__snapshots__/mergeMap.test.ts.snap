// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`I <3 RxJS snapshot 1`] = `
Array [
  Object {
    "frame": 0,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "a1",
    },
  },
  Object {
    "frame": 1,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "b1",
    },
  },
  Object {
    "frame": 3,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "a2",
    },
  },
  Object {
    "frame": 4,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "b2",
    },
  },
  Object {
    "frame": 6,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "a3",
    },
  },
  Object {
    "frame": 7,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "b3",
    },
  },
  Object {
    "frame": 9,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "c1",
    },
  },
  Object {
    "frame": 12,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "c2",
    },
  },
  Object {
    "frame": 15,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "c3",
    },
  },
  Object {
    "frame": 18,
    "notification": Notification {
      "error": undefined,
      "hasValue": false,
      "kind": "C",
      "value": undefined,
    },
  },
]
`;

exports[`mergeMap snapshot 1`] = `
Array [
  Object {
    "frame": 0,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "a1",
    },
  },
  Object {
    "frame": 1,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "b1",
    },
  },
  Object {
    "frame": 3,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "a2",
    },
  },
  Object {
    "frame": 4,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "b2",
    },
  },
  Object {
    "frame": 6,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "a3",
    },
  },
  Object {
    "frame": 7,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "b3",
    },
  },
  Object {
    "frame": 9,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "c1",
    },
  },
  Object {
    "frame": 12,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "c2",
    },
  },
  Object {
    "frame": 15,
    "notification": Notification {
      "error": undefined,
      "hasValue": true,
      "kind": "N",
      "value": "c3",
    },
  },
  Object {
    "frame": 18,
    "notification": Notification {
      "error": undefined,
      "hasValue": false,
      "kind": "C",
      "value": undefined,
    },
  },
]
`;
